drivers/hmis/app/models/hmis/workflow_execution/engine.rb

```diff
@@ -0,0 +1,151 @@
+
+      next_task_steps(step).each(&:disable!)
+      log_event('undo_complete_step', user: user, step: step)
+      step.undo_complete_step!
+    end
+
+    def may_undo_complete_step?(step)
+      step.may_undo_complete_step?
+      next_task_steps(step).all?(&:may_disable?)
+    end
+
+    # def cancel_step!(step)
+    #   step.cancel!
+    # end
+
+    protected
+
+    # get all tasks nodes under step but treat those task nodes as leaves and stop searching (bounded depth-first search)
+    def next_task_steps(step)
+      nodes = template.graph.walk(entrypoint_ids: [step.node_id], stop_when: lambda(&:task?))
+      steps_by_node_id = instance.steps.index_by(&:node_id)
+      nodes.map { |node| steps_by_node_id[node.id] }.compact
+    end
+
+    def send_message(...)
+      message = Hmis::WorkflowExecution::Message.new(...)
+      message_handler.call(message)
+    end
+
+    def process_triggers(node, event_type)
+      node.triggers.each do |trigger|
+        next unless event_type == trigger.event
+
+        send_message(
+          type: trigger.message,
+          params: trigger.params,
+        )
+        log_event('message_sent', event_data: trigger.to_h)
+      end
+    end
+
+    def traverse_node(node)
+      return if node.endpoint?
+
+      if node.join_inflows?
+        return unless node.inflows.all? { |flow| evaluate_condition(flow.condition) }
+      end
+
+      outflows = node.outflows.sort_by(&:position)
+      if node.exclusive_outflows?
+        outflows = Array(outflows.detect { |flow| evaluate_condition(flow.condition) })
+      else
+        outflows = outflows.filter { |flow| evaluate_condition(flow.condition) }
+      end
+
+      raise "Node (#{node.id}) blocks flow. It should have a default outflow" if outflows.empty?
+
+      process_triggers(node, 'pass_gateway') if node.gateway?
+      outflows.each { |flow| visit_node(flow.target_node) }
+    end
+
+    def visit_node(node)
+      case node
+      when Hmis::WorkflowDefinition::Task
+        step = instance.steps.new(node: node)
+        step.enable!
+        assign_task!(step)
+      when Hmis::WorkflowDefinition::Gateway
+        traverse_node(node)
+      when Hmis::WorkflowDefinition::StartEvent
+        process_triggers(node, 'start_workflow')
+        traverse_node(node)
+      when Hmis::WorkflowDefinition::EndEvent
+        process_triggers(node, 'end_workflow')
+      else
+        raise "Got unhandled node #{node.class.name}"
+      end
+    end
+
+    def assign_task!(step)
+      assignment_handler.call(step.node).each do |user|
+        step.assignments.create!(user: user)
+      end
+    end
+
+    def evaluate_condition(expression)
+      # empty expression defaults to true
+      return true if expression.blank?
+
+      calculator = Dentaku::Calculator.new
+      defaults = calculator.dependencies(expression).to_h { |k| [k.to_sym, nil] }
+      calculator.evaluate!(expression, **defaults.merge(all_submitted_values))
```

Comment:
Is this where the `submitted_values` get "processed" (e.g. to determine if the responses constitute a referral rejection)? Can you give an example of what an expression looks like? thanks!